File: lib/models/encryption_config.dart
enum EncryptionAlgorithm {
  // Block Ciphers - Sorted by encryption strength (strongest first)
  
  // Tier 1: Maximum Security (256+ bit keys, modern design)
  threefish1024('Threefish-1024', 'Threefish Block Cipher (1024-bit)'),
  threefish512('Threefish-512', 'Threefish Block Cipher (512-bit)'),
  shacal2('SHACAL-2', 'SHACAL-2 Block Cipher (SHA-based)'),
  
  // Tier 2: High Security (256-bit capable, proven algorithms)
  aes('AES', 'Advanced Encryption Standard (Rijndael)'),
  serpent('Serpent', 'Serpent Block Cipher'),
  twofish('Twofish', 'Twofish Block Cipher'),
  camellia('Camellia', 'Camellia Block Cipher'),
  aria('ARIA', 'ARIA Block Cipher (Korean Standard)'),
  threefish256('Threefish-256', 'Threefish Block Cipher (256-bit)'),
  
  // Stream Ciphers - High Security
  chacha20('ChaCha20', 'ChaCha20 Stream Cipher'),
  salsa20('Salsa20', 'Salsa20 Stream Cipher'),
  xsalsa20('XSalsa20', 'XSalsa20 Stream Cipher'),
  hc256('HC-256', 'HC-256 Stream Cipher'),
  
  // Tier 3: Good Security (192/256-bit capable)
  cast256('CAST-256', 'CAST-256 Block Cipher'),
  mars('MARS', 'MARS Block Cipher'),
  rc6('RC6', 'RC6 Block Cipher'),
  
  // Stream Ciphers - Good Security  
  chacha12('ChaCha12', 'ChaCha12 Stream Cipher'),
  chacha8('ChaCha8', 'ChaCha8 Stream Cipher'),
  hc128('HC-128', 'HC-128 Stream Cipher'),
  rabbit('Rabbit', 'Rabbit Stream Cipher'),
  sosemanuk('SOSEMANUK', 'SOSEMANUK Stream Cipher'),
  
  // Tier 4: Standard Security (128-bit minimum)
  seed('SEED', 'SEED Block Cipher (Korean Standard)'),
  idea('IDEA', 'International Data Encryption Algorithm'),
  cast128('CAST-128', 'CAST-128 Block Cipher'),
  tea('TEA', 'Tiny Encryption Algorithm'),
  xtea('XTEA', 'Extended TEA'),
  rc5('RC5', 'RC5 Block Cipher'),
  
  // Stream Ciphers - Standard Security
  seal('SEAL', 'SEAL Stream Cipher'),
  wake('WAKE-OFB', 'WAKE Output Feedback Mode'),
  
  // Tier 5: Variable Key/Legacy Strong
  blowfish('Blowfish', 'Blowfish Block Cipher (Variable Key)'),
  rc2('RC2', 'RC2 Block Cipher (Variable Key)'),
  
  // Tier 6: Legacy/Weak (for compatibility only)
  tripleDes('3DES', 'Triple Data Encryption Standard'),
  skipjack('SKIPJACK', 'SKIPJACK Block Cipher'),
  des('DES', 'Data Encryption Standard (Legacy)'),
  rc4('RC4', 'RC4 Stream Cipher (Legacy)');

  const EncryptionAlgorithm(this.displayName, this.description);

  final String displayName;
  final String description;

  List<int> getSupportedKeySizes() {
    switch (this) {
      // Threefish variants
      case EncryptionAlgorithm.threefish1024:
        return [1024];
      case EncryptionAlgorithm.threefish512:
        return [512];
      case EncryptionAlgorithm.threefish256:
        return [256];
      
      // SHACAL-2
      case EncryptionAlgorithm.shacal2:
        return [128, 192, 256, 384, 512];
      
      // Standard 128/192/256-bit algorithms
      case EncryptionAlgorithm.aes:
      case EncryptionAlgorithm.serpent:
      case EncryptionAlgorithm.twofish:
      case EncryptionAlgorithm.camellia:
      case EncryptionAlgorithm.aria:
      case EncryptionAlgorithm.cast256:
      case EncryptionAlgorithm.mars:
      case EncryptionAlgorithm.rc6:
        return [128, 192, 256];
      
      // Stream ciphers - 256-bit
      case EncryptionAlgorithm.chacha20:
      case EncryptionAlgorithm.chacha12:
      case EncryptionAlgorithm.chacha8:
      case EncryptionAlgorithm.salsa20:
      case EncryptionAlgorithm.xsalsa20:
      case EncryptionAlgorithm.hc256:
        return [256];
      
      // Stream ciphers - 128-bit
      case EncryptionAlgorithm.hc128:
      case EncryptionAlgorithm.rabbit:
      case EncryptionAlgorithm.sosemanuk:
        return [128];
      
      // 128-bit only algorithms
      case EncryptionAlgorithm.seed:
      case EncryptionAlgorithm.idea:
      case EncryptionAlgorithm.cast128:
      case EncryptionAlgorithm.tea:
      case EncryptionAlgorithm.xtea:
        return [128];
      
      // SEAL stream cipher
      case EncryptionAlgorithm.seal:
        return [160];
      
      // WAKE stream cipher
      case EncryptionAlgorithm.wake:
        return [128, 256];
      
      // Variable key algorithms
      case EncryptionAlgorithm.blowfish:
        return [32, 64, 128, 256, 448];
      case EncryptionAlgorithm.rc5:
        return [64, 128, 192, 256];
      case EncryptionAlgorithm.rc2:
        return [40, 64, 128];
      case EncryptionAlgorithm.rc4:
        return [40, 56, 64, 128, 256];
      
      // Legacy algorithms
      case EncryptionAlgorithm.tripleDes:
        return [168]; // 3 × 56-bit keys
      case EncryptionAlgorithm.skipjack:
        return [80];
      case EncryptionAlgorithm.des:
        return [56];
    }
  }

  List<OperationMode> getSupportedModes() {
    switch (this) {
      // Stream ciphers - only support stream mode (using CTR as representation)
      case EncryptionAlgorithm.chacha20:
      case EncryptionAlgorithm.chacha12:
      case EncryptionAlgorithm.chacha8:
      case EncryptionAlgorithm.salsa20:
      case EncryptionAlgorithm.xsalsa20:
      case EncryptionAlgorithm.hc128:
      case EncryptionAlgorithm.hc256:
      case EncryptionAlgorithm.rabbit:
      case EncryptionAlgorithm.sosemanuk:
      case EncryptionAlgorithm.seal:
      case EncryptionAlgorithm.wake:
      case EncryptionAlgorithm.rc4:
        return [OperationMode.ctr]; // Stream ciphers work like CTR mode
      
      // Block ciphers with full mode support
      case EncryptionAlgorithm.aes:
      case EncryptionAlgorithm.serpent:
      case EncryptionAlgorithm.twofish:
      case EncryptionAlgorithm.camellia:
      case EncryptionAlgorithm.aria:
        return [
          OperationMode.gcm,  // Authenticated encryption (preferred)
          OperationMode.cbc,  // Traditional
          OperationMode.ctr,  // Stream-like
          OperationMode.cfb,  // Cipher feedback
          OperationMode.ofb,  // Output feedback
        ];
      
      // Block ciphers with standard modes (no GCM)
      case EncryptionAlgorithm.threefish1024:
      case EncryptionAlgorithm.threefish512:
      case EncryptionAlgorithm.threefish256:
      case EncryptionAlgorithm.shacal2:
      case EncryptionAlgorithm.cast256:
      case EncryptionAlgorithm.mars:
      case EncryptionAlgorithm.rc6:
      case EncryptionAlgorithm.seed:
      case EncryptionAlgorithm.idea:
      case EncryptionAlgorithm.cast128:
      case EncryptionAlgorithm.tea:
      case EncryptionAlgorithm.xtea:
      case EncryptionAlgorithm.blowfish:
      case EncryptionAlgorithm.rc5:
      case EncryptionAlgorithm.rc2:
        return [
          OperationMode.cbc,  // Traditional
          OperationMode.ctr,  // Stream-like
          OperationMode.cfb,  // Cipher feedback
          OperationMode.ofb,  // Output feedback
          OperationMode.ecb,  // Simple (less secure)
        ];
      
      // Legacy algorithms with basic modes
      case EncryptionAlgorithm.tripleDes:
      case EncryptionAlgorithm.des:
      case EncryptionAlgorithm.skipjack:
        return [
          OperationMode.cbc,
          OperationMode.ecb,
        ];
    }
  }
}

enum OperationMode {
  cbc('CBC', 'Cipher Block Chaining'),
  gcm('GCM', 'Galois/Counter Mode'),
  ecb('ECB', 'Electronic Codebook'),
  cfb('CFB', 'Cipher Feedback'),
  ofb('OFB', 'Output Feedback'),
  ctr('CTR', 'Counter Mode');

  const OperationMode(this.displayName, this.description);

  final String displayName;
  final String description;
}

class EncryptionConfig {
  final EncryptionAlgorithm algorithm;
  final int keySize;
  final OperationMode mode;
  final String password;
  final int threadCount;

  const EncryptionConfig({
    required this.algorithm,
    required this.keySize,
    required this.mode,
    required this.password,
    required this.threadCount,
  });

  EncryptionConfig copyWith({
    EncryptionAlgorithm? algorithm,
    int? keySize,
    OperationMode? mode,
    String? password,
    int? threadCount,
  }) {
    return EncryptionConfig(
      algorithm: algorithm ?? this.algorithm,
      keySize: keySize ?? this.keySize,
      mode: mode ?? this.mode,
      password: password ?? this.password,
      threadCount: threadCount ?? this.threadCount,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'algorithm': algorithm.name,
      'keySize': keySize,
      'mode': mode.name,
      'password': password,
      'threadCount': threadCount,
    };
  }

  factory EncryptionConfig.fromJson(Map<String, dynamic> json) {
    return EncryptionConfig(
      algorithm: EncryptionAlgorithm.values.firstWhere(
        (e) => e.name == json['algorithm'],
      ),
      keySize: json['keySize'],
      mode: OperationMode.values.firstWhere(
        (e) => e.name == json['mode'],
      ),
      password: json['password'],
      threadCount: json['threadCount'],
    );
  }

  @override
  String toString() {
    return 'EncryptionConfig(algorithm: ${algorithm.displayName}, keySize: $keySize, mode: ${mode.displayName}, threadCount: $threadCount)';
  }
}

// Default configuration
class EncryptionDefaults {
  static const EncryptionAlgorithm defaultAlgorithm = EncryptionAlgorithm.aes;
  static const int defaultKeySize = 256;
  static const OperationMode defaultMode = OperationMode.gcm;
  static const int defaultThreadCount = 4;
  static const String defaultPassword = '';

  static EncryptionConfig get defaultConfig => const EncryptionConfig(
    algorithm: defaultAlgorithm,
    keySize: defaultKeySize,
    mode: defaultMode,
    password: defaultPassword,
    threadCount: defaultThreadCount,
  );
}

File: android/app/src/main/kotlin/com/binah/cryptingtool/models/EncryptionConfig.kt
package com.binah.cryptingtool.models

import android.os.Parcelable
import kotlinx.parcelize.Parcelize

enum class EncryptionAlgorithm(val displayName: String, val description: String) {
    // Tier 1: Maximum Security - Modern & Post-Quantum Ready
    AES256("AES-256", "Advanced Encryption Standard - 256-bit (Maximum Security)"),
    SERPENT256("Serpent-256", "Serpent Block Cipher - 256-bit (Maximum Security)"),
    TWOFISH256("Twofish-256", "Twofish Block Cipher - 256-bit (Maximum Security)"),
    
    // Tier 2: High Security - Modern Algorithms
    AES192("AES-192", "Advanced Encryption Standard - 192-bit (High Security)"),
    AES128("AES-128", "Advanced Encryption Standard - 128-bit (High Security)"),
    SERPENT192("Serpent-192", "Serpent Block Cipher - 192-bit (High Security)"),
    SERPENT128("Serpent-128", "Serpent Block Cipher - 128-bit (High Security)"),
    TWOFISH192("Twofish-192", "Twofish Block Cipher - 192-bit (High Security)"),
    TWOFISH128("Twofish-128", "Twofish Block Cipher - 128-bit (High Security)"),
    
    // Tier 3: Strong Security - AES Finalists & Modern Ciphers
    RC6("RC6", "Rivest Cipher 6 - Variable Key Length"),
    MARS("MARS", "IBM MARS Block Cipher"),
    RC5("RC5", "Rivest Cipher 5 - Variable Parameters"),
    SKIPJACK("Skipjack", "NSA Skipjack Block Cipher"),
    
    // Tier 4: Reliable Security - Established Algorithms
    BLOWFISH("Blowfish", "Blowfish Block Cipher - Variable Key (32-448 bits)"),
    CAST128("CAST-128", "CAST-128 Block Cipher (128-bit key)"),
    CAST256("CAST-256", "CAST-256 Block Cipher (128/160/192/224/256-bit key)"),
    CAMELLIA("Camellia", "NTT/Mitsubishi Camellia Block Cipher"),
    
    // Tier 5: Stream Ciphers - High Performance
    CHACHA20("ChaCha20", "ChaCha20 Stream Cipher (256-bit key)"),
    SALSA20("Salsa20", "Salsa20 Stream Cipher (128/256-bit key)"),
    XSALSA20("XSalsa20", "Extended Salsa20 Stream Cipher (256-bit key)"),
    HC128("HC-128", "HC-128 Stream Cipher (128-bit key)"),
    HC256("HC-256", "HC-256 Stream Cipher (256-bit key)"),
    RABBIT("Rabbit", "Rabbit Stream Cipher (128-bit key)"),
    SOSEMANUK("Sosemanuk", "Sosemanuk Stream Cipher (128-256 bit key)"),
    
    // Tier 6: Specialized & National Algorithms
    ARIA("ARIA", "Korean ARIA Block Cipher (128/192/256-bit)"),
    SEED("SEED", "Korean SEED Block Cipher (128-bit key)"),
    SM4("SM4", "Chinese SM4 Block Cipher (128-bit key)"),
    GOST28147("GOST 28147-89", "Russian GOST Block Cipher (256-bit key)"),
    
    // Tier 7: Legacy Strong Algorithms
    DES3("3DES", "Triple Data Encryption Standard (168-bit effective)"),
    IDEA("IDEA", "International Data Encryption Algorithm (128-bit)"),
    RC2("RC2", "Rivest Cipher 2 - Variable Key Length"),
    SAFER("SAFER", "Secure And Fast Encryption Routine"),
    SAFER_PLUS("SAFER+", "Enhanced SAFER Block Cipher"),
    
    // Tier 8: Historical & Compatibility
    DES("DES", "Data Encryption Standard (56-bit key) - Legacy Only"),
    RC4("RC4", "Rivest Cipher 4 Stream Cipher - Legacy");
    
    fun getSupportedKeySizes(): List<Int> {
        return when (this) {
            // AES Family
            AES256 -> listOf(256)
            AES192 -> listOf(192)
            AES128 -> listOf(128)
            
            // Serpent Family
            SERPENT256 -> listOf(256)
            SERPENT192 -> listOf(192)
            SERPENT128 -> listOf(128)
            
            // Twofish Family
            TWOFISH256 -> listOf(256)
            TWOFISH192 -> listOf(192)
            TWOFISH128 -> listOf(128)
            
            // Stream Ciphers
            CHACHA20, XSALSA20, HC256, SOSEMANUK, GOST28147 -> listOf(256)
            SALSA20 -> listOf(128, 256)
            HC128, RABBIT -> listOf(128)
            
            // Variable key algorithms
            RC6, MARS -> listOf(128, 192, 256)
            RC5 -> listOf(128, 160, 192, 224, 256)
            BLOWFISH -> listOf(128, 192, 256, 448)
            CAST256 -> listOf(128, 160, 192, 224, 256)
            RC2 -> listOf(40, 64, 128)
            
            // Fixed key algorithms
            SKIPJACK, CAST128, CAMELLIA, ARIA, SEED, SM4, IDEA -> listOf(128)
            DES3 -> listOf(168)
            DES -> listOf(56)
            RC4 -> listOf(40, 128)
            SAFER, SAFER_PLUS -> listOf(128, 160)
        }
    }
    
    fun getSupportedModes(): List<OperationMode> {
        // Stream ciphers only support stream mode (represented as CTR)
        if (isStreamCipher()) {
            return listOf(OperationMode.CTR)
        }
        
        // Block ciphers support various modes
        return when (this) {
            // Modern block ciphers - full mode support
            AES256, AES192, AES128, SERPENT256, SERPENT192, SERPENT128,
            TWOFISH256, TWOFISH192, TWOFISH128, CAMELLIA, ARIA -> listOf(
                OperationMode.GCM,    // Authenticated encryption (preferred)
                OperationMode.CBC,    // Standard mode
                OperationMode.CFB,    // Cipher feedback
                OperationMode.OFB,    // Output feedback
                OperationMode.CTR,    // Counter mode
                OperationMode.ECB     // Electronic codebook (less secure)
            )
            
            // Good block ciphers - most modes
            RC6, MARS, CAST256, SEED, SM4, IDEA, BLOWFISH -> listOf(
                OperationMode.CBC,
                OperationMode.CFB,
                OperationMode.OFB,
                OperationMode.CTR,
                OperationMode.ECB
            )
            
            // Older/simpler block ciphers - basic modes
            CAST128, SKIPJACK, RC5, DES3, DES, RC2, SAFER, SAFER_PLUS -> listOf(
                OperationMode.CBC,
                OperationMode.CFB,
                OperationMode.OFB,
                OperationMode.ECB
            )
            
            // Research/specialized algorithms - limited modes
            GOST28147 -> listOf(
                OperationMode.CBC,
                OperationMode.ECB
            )
            
            else -> listOf(OperationMode.CBC, OperationMode.ECB)
        }
    }
    
    private fun isStreamCipher(): Boolean {
        return this in listOf(
            CHACHA20, SALSA20, XSALSA20, HC128, HC256,
            RABBIT, SOSEMANUK, RC4
        )
    }
}

enum class OperationMode(val displayName: String, val description: String) {
    CBC("CBC", "Cipher Block Chaining"),
    GCM("GCM", "Galois/Counter Mode"),
    ECB("ECB", "Electronic Codebook"),
    CFB("CFB", "Cipher Feedback"),
    OFB("OFB", "Output Feedback"),
    CTR("CTR", "Counter Mode")
}

@Parcelize
data class EncryptionConfig(
    val algorithm: EncryptionAlgorithm = EncryptionAlgorithm.AES256,
    val keySize: Int = 256,
    val mode: OperationMode = OperationMode.GCM,
    val password: String = "",
    val threadCount: Int = 4
) : Parcelable {
    
    companion object {
        val DEFAULT = EncryptionConfig()
    }
    
    fun copy(
        algorithm: EncryptionAlgorithm = this.algorithm,
        keySize: Int = this.keySize,
        mode: OperationMode = this.mode,
        password: String = this.password,
        threadCount: Int = this.threadCount
    ) = EncryptionConfig(algorithm, keySize, mode, password, threadCount)
}